feat: implement route authentication middleware

CMI-James opened the pull request:
<p align="center"> <img src="https://github.com/user-attachments/assets/25cfd5b8-e303-4af6-920e-e42d1ac8d418" alt="CLR-S (2)"> </p>

# Pull Request | GrantChain

## 1. Issue Link

- Closes #118

---

## 2. Brief Description of the Issue

Implement server-side route authentication middleware to protect dashboard routes. Previously relied only on client-side auth checks, lacking proper server-side protection.

---

## 3. Type of Change

- [x] �� Enhancement (non-breaking change which adds functionality)

---

## 4. Changes Made

- **Created `middleware.ts`** with Supabase session validation
- **Protected all `/dashboard/*` routes** from unauthenticated access
- **Added special handling** for public profile routes
- **Removed redundant client-side auth checks** from dashboard layout
- **Preserved redirect destination** for post-login redirects
- **Maintained OAuth compatibility** and role-based access

---

## 7. Important Notes

- **Server-side security** at middleware level
- **No breaking changes** - existing auth flows work seamlessly
- **Performance optimized** with efficient session checking
- **Public profiles** remain accessible without full auth

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Added an authentication gate for dashboard routes, requiring a valid session.
  * Unauthenticated users are redirected to the login page with a return path.
  * /dashboard/public-profile remains accessible without signing in.
  * Static assets, API and auth routes are excluded from the gate.
  * Authentication errors are handled by redirecting users to login to preserve access continuity.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

coxmars commented on src/app/dashboard/layout.tsx:
I noticed you removed the wallet validation logic from the layout that was checking for address and redirecting when no wallet was connected, can you explain me this? Also remember to add a loom video showing this working as expected.

CMI-James replied:
Apologies for the removal, it was an oversight. I will revert, update and create a loom record as requested.

coxmars replied:
> Apologies for the removal, it was an oversight. I will revert, update and create a loom record as requested.

Perfect, I will be waiting for that

CMI-James committed:
feat: implement route authentication middleware for dashboard protection

- Add server-side authentication middleware using @supabase/ssr
- Protect all /dashboard routes with session validation
- Redirect unauthenticated users to /login with redirectTo parameter
- Move middleware from root to src/ directory for Next.js 15 compatibility
- Sync with upstream GrantChain/GrantFox develop branch
- Remove empty root middleware.ts file
- Maintain compatibility with existing OAuth flows and auth context
- Replace forEach with for...of loops for better performance

CMI-James commented:
@coxmars here is a loom video for the implementation. 
https://www.loom.com/share/cc6cb2021fee47338c86809d19d070f4?sid=af5b4bac-c15b-4bce-bae0-d955562a988b

CMI-James committed:
fix: address critical security and UX issues in middleware

- Preserve query parameters in redirectTo for post-login navigation
- Allow public profile routes without authentication
- Use getUser() instead of getSession() for secure server-side validation
- Prevent session spoofing by validating tokens server-side

Security improvements:
- Critical: Switch from getSession() to getUser() to prevent auth bypass
- Critical: Public profiles now accessible without authentication
- Minor: Query parameters preserved in redirect URLs

CMI-James committed:
refactor: optimize middleware for production and improve performance

- Remove log spam by limiting debug logs to development only
- Remove unreachable code paths (matcher already filters routes)
- Move public profile exclusion to matcher for better performance
- Add environment variable validation with fail-fast behavior
- Simplify logic by relying on matcher constraints

Performance improvements:
- Public profiles bypass middleware entirely (no execution overhead)
- Reduced conditional branches for faster execution
- Clear error handling for missing env variables

CMI-James committed:
feat: update middleware authentication logic
